From 4c07842203eb51751b8ae9acb9d3fdd88521aa44 Mon Sep 17 00:00:00 2001
From: Yi Zhao <yi.zhao@windriver.com>
Date: Thu, 13 Jan 2022 11:29:48 +0800
Subject: [PATCH] Add the WRLinux LTS21 product

The product has been created using WRLinux1019 as a basis, and making
sure that the build works.

Upstream-Status: Pending

Signed-off-by: Yi Zhao <yi.zhao@windriver.com>
---
 CMakeLists.txt                                |   5 +
 build_product                                 |   1 +
 products/wrlinuxlts21/CMakeLists.txt          |   6 +
 products/wrlinuxlts21/overlays/.gitkeep       |   0
 products/wrlinuxlts21/product.yml             |  19 ++
 .../profiles/basic-embedded.profile           |  30 +++
 .../profiles/draft_stig_wrlinux_disa.profile  | 230 ++++++++++++++++++
 .../wrlinuxlts21/transforms/constants.xslt    |  12 +
 .../oval/installed_OS_is_wrlinuxlts21.xml     |  49 ++++
 ssg/build_ovals.py                            |   5 +
 ssg/constants.py                              |   5 +-
 11 files changed, 360 insertions(+), 2 deletions(-)
 create mode 100644 products/wrlinuxlts21/CMakeLists.txt
 create mode 100644 products/wrlinuxlts21/overlays/.gitkeep
 create mode 100644 products/wrlinuxlts21/product.yml
 create mode 100644 products/wrlinuxlts21/profiles/basic-embedded.profile
 create mode 100644 products/wrlinuxlts21/profiles/draft_stig_wrlinux_disa.profile
 create mode 100644 products/wrlinuxlts21/transforms/constants.xslt
 create mode 100644 shared/checks/oval/installed_OS_is_wrlinuxlts21.xml

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7225a766e8..d9acb3903f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -96,6 +96,7 @@ option(SSG_PRODUCT_UBUNTU2004 "If enabled, the Ubuntu 20.04 SCAP content will be
 option(SSG_PRODUCT_VSEL "If enabled, the McAfee VSEL SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
 option(SSG_PRODUCT_WRLINUX8 "If enabled, the WRLinux8 SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
 option(SSG_PRODUCT_WRLINUX1019 "If enabled, the WRLinux1019 SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
+option(SSG_PRODUCT_WRLINUXLTS21 "If enabled, the WRLinuxLTS21 SCAP content will be built" ${SSG_PRODUCT_DEFAULT})
 
 option(SSG_CENTOS_DERIVATIVES_ENABLED "If enabled, CentOS derivative content will be built from the RHEL content" TRUE)
 option(SSG_SCIENTIFIC_LINUX_DERIVATIVES_ENABLED "If enabled, Scientific Linux derivative content will be built from the RHEL content" TRUE)
@@ -293,6 +294,7 @@ message(STATUS "Ubuntu 20.04: ${SSG_PRODUCT_UBUNTU2004}")
 message(STATUS "McAfee VSEL: ${SSG_PRODUCT_VSEL}")
 message(STATUS "WRLinux 8: ${SSG_PRODUCT_WRLINUX8}")
 message(STATUS "WRLinux 1019: ${SSG_PRODUCT_WRLINUX1019}")
+message(STATUS "WRLinux LTS21: ${SSG_PRODUCT_WRLINUXLTS21}")
 
 
 
@@ -416,6 +418,9 @@ endif()
 if (SSG_PRODUCT_WRLINUX1019)
     add_subdirectory("products/wrlinux1019" "wrlinux1019")
 endif()
+if (SSG_PRODUCT_WRLINUXLTS21)
+    add_subdirectory("products/wrlinuxlts21" "wrlinuxlts21")
+endif()
 
 # ZIP only contains source datastreams and kickstarts, people who
 # want sources to build from should get the tarball instead.
diff --git a/build_product b/build_product
index 4d7aae8762..c0d4fe7772 100755
--- a/build_product
+++ b/build_product
@@ -310,6 +310,7 @@ all_cmake_products=(
 	VSEL
 	WRLINUX8
 	WRLINUX1019
+	WRLINUXLTS21
 	MACOS1015
 )
 
diff --git a/products/wrlinuxlts21/CMakeLists.txt b/products/wrlinuxlts21/CMakeLists.txt
new file mode 100644
index 0000000000..e871d3031f
--- /dev/null
+++ b/products/wrlinuxlts21/CMakeLists.txt
@@ -0,0 +1,6 @@
+# Sometimes our users will try to do: "cd wrlinuxlts21; cmake ." That needs to error in a nice way.
+if ("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
+    message(FATAL_ERROR "cmake has to be used on the root CMakeLists.txt, see the Building ComplianceAsCode section in the Developer Guide!")
+endif()
+
+ssg_build_product("wrlinuxlts21")
diff --git a/products/wrlinuxlts21/overlays/.gitkeep b/products/wrlinuxlts21/overlays/.gitkeep
new file mode 100644
index 0000000000..e69de29bb2
diff --git a/products/wrlinuxlts21/product.yml b/products/wrlinuxlts21/product.yml
new file mode 100644
index 0000000000..fcf69ce3c6
--- /dev/null
+++ b/products/wrlinuxlts21/product.yml
@@ -0,0 +1,19 @@
+product: wrlinuxlts21
+full_name: WRLinux LTS21
+type: platform
+
+benchmark_id: WRLINUX
+benchmark_root: "../../linux_os/guide"
+
+profiles_root: "./profiles"
+
+pkg_manager: "dnf"
+
+init_system: "systemd"
+
+cpes_root: "../../shared/applicability"
+cpes:
+  - wrlinuxlts21:
+      name: "cpe:/o:windriver:wrlinux:lts21"
+      title: "Wind River Linux LTS21"
+      check_id: installed_OS_is_wrlinuxlts21
diff --git a/products/wrlinuxlts21/profiles/basic-embedded.profile b/products/wrlinuxlts21/profiles/basic-embedded.profile
new file mode 100644
index 0000000000..7bc1c4d867
--- /dev/null
+++ b/products/wrlinuxlts21/profiles/basic-embedded.profile
@@ -0,0 +1,30 @@
+documentation_complete: true
+
+title: 'Basic Profile for Embedded Systems'
+
+description: |-
+    This profile contains items common to many embedded Linux installations.
+    Regardless of your system's deployment objective, all of these checks should pass.
+
+selections:
+    - sshd_idle_timeout_value=10_minutes
+    - var_accounts_fail_delay=4
+    - var_accounts_passwords_pam_faillock_deny=3
+    - var_accounts_passwords_pam_faillock_unlock_time=604800
+    - var_accounts_user_umask=077
+    - var_auditd_action_mail_acct=root
+    - var_auditd_max_log_file_action=rotate
+    - var_auditd_num_logs=5
+    - var_auditd_space_left_action=email
+    - var_selinux_policy_name=targeted
+    - accounts_logon_fail_delay
+    - accounts_max_concurrent_login_sessions
+    - accounts_no_uid_except_zero
+    - accounts_password_all_shadowed
+    - accounts_root_path_dirs_no_write
+    - accounts_umask_etc_login_defs
+    - accounts_umask_etc_profile
+    - file_permissions_home_dirs
+    - no_empty_passwords
+    - no_netrc_files
+    - root_path_no_dot
diff --git a/products/wrlinuxlts21/profiles/draft_stig_wrlinux_disa.profile b/products/wrlinuxlts21/profiles/draft_stig_wrlinux_disa.profile
new file mode 100644
index 0000000000..f8a32037ce
--- /dev/null
+++ b/products/wrlinuxlts21/profiles/draft_stig_wrlinux_disa.profile
@@ -0,0 +1,230 @@
+documentation_complete: true
+
+title: 'DRAFT DISA STIG for Wind River Linux'
+
+description: |-
+    This profile contains configuration checks that align to the
+    DISA STIG for Wind River Linux.
+    This profile is being developed under the DoD consensus model to
+    become a STIG in coordination with DISA FSO.
+    What is the status of the Wind River Linux STIG?
+    The Wind River Linux STIG is in development under the DoD consensus model
+    and Wind River has started the process to get approval from DISA. However, in
+    the absence of an approved SRG or STIG, vendor recommendations may be used
+    instead. The current contents constitute the vendor recommendations at the
+    time of the product release containing these contents.
+    Note that changes are expected before approval is granted, and those changes
+    will be made available in future Wind River Linux LTS21 RCPL releases.
+    More information, including the following, is available from the DISA FAQs
+    at https://public.cyber.mil/stigs/faqs/
+
+selections:
+    - var_accounts_max_concurrent_login_sessions=10
+    - var_accounts_minimum_age_login_defs=1
+    - var_accounts_maximum_age_login_defs=60
+    - sysctl_net_ipv4_conf_all_accept_source_route_value=disabled
+    - sysctl_net_ipv4_conf_default_accept_source_route_value=disabled
+    - sysctl_net_ipv4_icmp_echo_ignore_broadcasts_value=enabled
+    - sysctl_net_ipv4_conf_all_accept_redirects_value=disabled
+    - sysctl_net_ipv6_conf_all_accept_source_route_value=disabled
+    - var_account_disable_post_pw_expiration=0
+    - var_accounts_tmout=10_min
+    - var_selinux_state=enforcing
+    - var_accounts_passwords_pam_faillock_deny=3
+    - banner_etc_issue
+    - sshd_enable_warning_banner
+    - accounts_max_concurrent_login_sessions
+    - package_screen_installed
+    - sshd_use_approved_ciphers
+    - grub2_enable_fips_mode
+    - audit_rules_privileged_commands_su
+    - audit_rules_privileged_commands_sudo
+    - audit_rules_sysadmin_actions
+    - audit_rules_privileged_commands_chsh
+    - audit_rules_privileged_commands_passwd
+    - audit_rules_privileged_commands_unix_chkpwd
+    - audit_rules_privileged_commands_gpasswd
+    - audit_rules_privileged_commands_chage
+    - audit_rules_privileged_commands_umount
+    - audit_rules_privileged_commands_postdrop
+    - audit_rules_privileged_commands_postqueue
+    - audit_rules_privileged_commands_ssh_keysign
+    - audit_rules_privileged_commands_crontab
+    - audit_rules_system_shutdown
+    - audit_rules_dac_modification_chown
+    - audit_rules_dac_modification_fchown
+    - audit_rules_dac_modification_lchown
+    - audit_rules_dac_modification_fchownat
+    - audit_rules_privileged_commands_pam_timestamp_check
+    - set_password_hashing_algorithm_logindefs
+    - set_password_hashing_algorithm_libuserconf
+    - sshd_allow_only_protocol2
+    - accounts_minimum_age_login_defs
+    - accounts_password_set_min_life_existing
+    - accounts_maximum_age_login_defs
+    - accounts_password_set_max_life_existing
+    - accounts_password_pam_unix_remember
+    - require_singleuser_auth
+    - accounts_logon_fail_delay
+    - no_empty_passwords
+    - disable_ctrlaltdel_reboot
+    - accounts_no_uid_except_zero
+    - accounts_user_interactive_home_directory_defined
+    - accounts_have_homedir_login_defs
+    - accounts_user_interactive_home_directory_exists
+    - file_permissions_home_directories
+    - file_ownership_home_directories
+    - file_groupownership_home_directories
+    - accounts_users_home_files_ownership
+    - accounts_users_home_files_groupownership
+    - accounts_users_home_files_permissions
+    - accounts_user_dot_user_ownership
+    - accounts_user_dot_group_ownership
+    - file_permission_user_init_files
+    - accounts_user_home_paths_only
+    - accounts_user_dot_no_world_writable_programs
+    - mount_option_home_nosuid
+    - mount_option_nosuid_removable_partitions
+    - mount_option_nosuid_remote_filesystems
+    - mount_option_noexec_remote_filesystems
+    - dir_perms_world_writable_system_owned
+    - rsyslog_cron_logging
+    - file_owner_cron_allow
+    - file_groupowner_cron_allow
+    - service_kdump_disabled
+    - partition_for_home
+    - partition_for_var
+    - partition_for_var_log_audit
+    - partition_for_tmp
+    - rsyslog_remote_loghost
+    - display_login_attempts
+    - sshd_print_last_log
+    - sshd_disable_root_login
+    - network_configure_name_resolution
+    - no_user_host_based_files
+    - no_host_based_files
+    - sshd_disable_rhosts_rsa
+    - sysctl_net_ipv4_conf_all_accept_source_route
+    - sysctl_net_ipv4_conf_default_accept_source_route
+    - sysctl_net_ipv4_icmp_echo_ignore_broadcasts
+    - sysctl_net_ipv4_conf_all_accept_redirects
+    - sysctl_net_ipv4_conf_default_send_redirects
+    - sysctl_net_ipv4_conf_all_send_redirects
+    - network_sniffer_disabled
+    - postfix_prevent_unrestricted_relay
+    - package_vsftpd_removed
+    - tftpd_uses_secure_mode
+    - sshd_enable_x11_forwarding
+    - snmpd_not_default_password
+    - file_permissions_sshd_pub_key
+    - file_permissions_sshd_private_key
+    - sshd_enable_strictmodes
+    - sshd_use_priv_separation
+    - sshd_disable_compression
+    - sysctl_net_ipv4_ip_forward
+    - mount_option_krb_sec_remote_filesystems
+    - sysctl_net_ipv6_conf_all_accept_source_route
+    - accounts_umask_etc_login_defs
+    - sshd_do_not_permit_user_env
+    - disable_host_auth
+    - package_rsh-server_removed
+    - package_ypserv_removed
+    - package_telnet-server_removed
+    - gid_passwd_group_same
+    - sshd_disable_empty_passwords
+    - kernel_module_usb-storage_disabled
+    - service_autofs_disabled
+    - account_disable_post_pw_expiration
+    - accounts_tmout
+    - sshd_set_idle_timeout
+    - var_sshd_set_keepalive=0
+    - sshd_set_keepalive_0
+    - sshd_use_approved_macs
+    - rpm_verify_permissions
+    - sshd_disable_gssapi_auth
+    - sshd_disable_kerb_auth
+    - auditd_audispd_configure_remote_server
+    - auditd_audispd_encrypt_sent_records
+    - auditd_audispd_disk_full_action
+    - auditd_audispd_network_failure_action
+    - auditd_data_retention_space_left
+    - auditd_data_retention_action_mail_acct
+    - chronyd_or_ntpd_set_maxpoll
+    - sudo_remove_nopasswd
+    - sudo_remove_no_authenticate
+    - audit_rules_privileged_commands
+    - selinux_user_login_roles
+    - service_sshd_enabled
+    - selinux_state
+    - selinux_policytype
+    - audit_rules_execution_semanage
+    - audit_rules_execution_setsebool
+    - audit_rules_execution_chcon
+    - service_auditd_enabled
+    - audit_rules_usergroup_modification_passwd
+    - audit_rules_usergroup_modification_group
+    - audit_rules_usergroup_modification_gshadow
+    - audit_rules_usergroup_modification_shadow
+    - audit_rules_usergroup_modification_opasswd
+    - audit_rules_unsuccessful_file_modification_creat
+    - audit_rules_unsuccessful_file_modification_open
+    - audit_rules_unsuccessful_file_modification_openat
+    - audit_rules_unsuccessful_file_modification_open_by_handle_at
+    - audit_rules_unsuccessful_file_modification_truncate
+    - audit_rules_unsuccessful_file_modification_ftruncate
+    - audit_rules_file_deletion_events_rename
+    - audit_rules_file_deletion_events_renameat
+    - audit_rules_file_deletion_events_rmdir
+    - audit_rules_file_deletion_events_unlink
+    - audit_rules_file_deletion_events_unlinkat
+    - audit_rules_kernel_module_loading_init
+    - audit_rules_kernel_module_loading_delete
+    - audit_rules_kernel_module_loading_finit
+    - configure_firewalld_rate_limiting
+    - audit_rules_login_events_faillock
+    - audit_rules_login_events_lastlog
+    - audit_rules_dac_modification_chmod
+    - audit_rules_dac_modification_fchmod
+    - audit_rules_dac_modification_fchmodat
+    - audit_rules_dac_modification_setxattr
+    - audit_rules_dac_modification_fsetxattr
+    - audit_rules_dac_modification_lsetxattr
+    - audit_rules_dac_modification_removexattr
+    - audit_rules_dac_modification_fremovexattr
+    - audit_rules_dac_modification_lremovexattr
+    - accounts_password_pam_retry
+    - accounts_password_pam_ucredit
+    - accounts_password_pam_lcredit
+    - accounts_password_pam_dcredit
+    - accounts_password_pam_difok
+    - accounts_password_pam_minclass
+    - accounts_password_pam_maxrepeat
+    - accounts_password_pam_maxclassrepeat
+    - set_password_hashing_algorithm_systemauth
+    - accounts_password_pam_minlen
+    - grub2_password
+    - grub2_uefi_password
+    - rpm_verify_hashes
+    - installed_OS_is_vendor_supported
+    - no_files_unowned_by_user
+    - file_permissions_ungroupowned
+    - selinux_all_devicefiles_labeled
+    - accounts_umask_interactive_users
+    - rsyslog_nolisten
+    - service_firewalld_enabled
+    - package_tftp-server_removed
+    - configure_firewalld_ports
+    - sssd_ldap_start_tls
+    - sssd_ldap_configure_tls_ca_dir
+    - sssd_ldap_configure_tls_ca
+    - accounts_password_pam_ocredit
+    - aide_periodic_cron_checking
+    - aide_scan_notification
+    - ensure_gpgcheck_local_packages
+    - grub2_no_removeable_media
+    - auditd_data_retention_admin_space_left_action
+    - kernel_module_dccp_disabled
+    - audit_rules_usergroup_modification
+    - accounts_passwords_pam_faillock_deny
+    - accounts_passwords_pam_faillock_deny_root
+    - package_openssh-server_installed
diff --git a/products/wrlinuxlts21/transforms/constants.xslt b/products/wrlinuxlts21/transforms/constants.xslt
new file mode 100644
index 0000000000..be025cb661
--- /dev/null
+++ b/products/wrlinuxlts21/transforms/constants.xslt
@@ -0,0 +1,12 @@
+<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
+
+<xsl:include href="../../../shared/transforms/shared_constants.xslt"/>
+
+<xsl:variable name="product_long_name">WRLinux</xsl:variable>
+<xsl:variable name="product_short_name">WRLinux</xsl:variable>
+<xsl:variable name="product_stig_id_name">empty</xsl:variable>
+<xsl:variable name="prod_type">wrlinux</xsl:variable>
+
+<xsl:variable name="cisuri">empty</xsl:variable>
+
+</xsl:stylesheet>
diff --git a/shared/checks/oval/installed_OS_is_wrlinuxlts21.xml b/shared/checks/oval/installed_OS_is_wrlinuxlts21.xml
new file mode 100644
index 0000000000..2cc58c500a
--- /dev/null
+++ b/shared/checks/oval/installed_OS_is_wrlinuxlts21.xml
@@ -0,0 +1,49 @@
+<def-group>
+  <definition class="inventory" id="installed_OS_is_wrlinuxlts21" version="1">
+    <metadata>
+      <title>WRLinux LTS21</title>
+      <affected family="unix">
+        <platform>multi_platform_all</platform>
+      </affected>
+      <reference ref_id="cpe:/o:windriver:wrlinux"
+      source="CPE" />
+      <description>The operating system installed on the system is
+      Wind River Linux LTS21</description>
+    </metadata>
+    <criteria operator="AND">
+      <criterion comment="Installed operating system is part of the unix family" test_ref="test_unix_wrlinuxlts21" />
+      <criterion comment="Produce name is WRLinux" test_ref="test_wrlinuxlts21_name" />
+      <criterion comment="WRLinux version is LTS21" test_ref="test_wrlinuxlts21_version" />
+    </criteria>
+  </definition>
+
+  <ind:family_test check="all" check_existence="at_least_one_exists" comment="installed OS part of unix family" id="test_unix_wrlinuxlts21" version="1">
+    <ind:object object_ref="obj_unix_wrlinuxlts21" />
+    <ind:state state_ref="state_unix_wrlinuxlts21" />
+  </ind:family_test>
+
+  <ind:family_state id="state_unix_wrlinuxlts21" version="1">
+    <ind:family>unix</ind:family>
+  </ind:family_state>
+
+  <ind:family_object id="obj_unix_wrlinuxlts21" version="1" />
+
+  <ind:textfilecontent54_test check="all" check_existence="at_least_one_exists" comment="Check WRLinux10.19 version" id="test_wrlinuxlts21_name" version="1">
+    <ind:object object_ref="obj_name_wrlinuxlts21" />
+  </ind:textfilecontent54_test>
+  <ind:textfilecontent54_object id="obj_name_wrlinuxlts21" version="1" comment="Check WRLinux LTS21 produce name">
+    <ind:filepath>/etc/os-release</ind:filepath>
+    <ind:pattern operation="pattern match">^NAME=.Wind[\s]+River[\s]+Linux.*$</ind:pattern>
+    <ind:instance datatype="int">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <ind:textfilecontent54_test check="all" check_existence="at_least_one_exists" comment="Check WRLinux LTS21 version" id="test_wrlinuxlts21_version" version="1">
+    <ind:object object_ref="obj_version_wrlinuxlts21" />
+  </ind:textfilecontent54_test>
+  <ind:textfilecontent54_object id="obj_version_wrlinuxlts21" version="1" comment="Check WRLinux LTS21 version">
+    <ind:filepath>/etc/os-release</ind:filepath>
+    <ind:pattern operation="pattern match">^VERSION=.10\.21.*$</ind:pattern>
+    <ind:instance datatype="int">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+</def-group>
diff --git a/ssg/build_ovals.py b/ssg/build_ovals.py
index d5409448fc..512c47d2f8 100644
--- a/ssg/build_ovals.py
+++ b/ssg/build_ovals.py
@@ -27,6 +27,11 @@ def _check_is_applicable_for_product(oval_check_def, product):
 
     product, product_version = utils.parse_name(product)
 
+    # Get th correct product name and product version for WRLinux LTS21
+    if product == "wrlinuxlts" and product_version == "21":
+        product = "wrlinux"
+        product_version = "LTS21"
+
     # Define general platforms
     multi_platforms = ['<platform>multi_platform_all',
                        '<platform>multi_platform_' + product]
diff --git a/ssg/constants.py b/ssg/constants.py
index 327afe31f9..4d1302aacd 100644
--- a/ssg/constants.py
+++ b/ssg/constants.py
@@ -55,7 +55,7 @@ product_directories = [
     'sle12', 'sle15',
     'ubuntu1604', 'ubuntu1804', 'ubuntu2004',
     'vsel',
-    'wrlinux8', 'wrlinux1019'
+    'wrlinux8', 'wrlinux1019', 'wrlinuxlts21'
 ]
 
 JINJA_MACROS_BASE_DEFINITIONS = os.path.join(os.path.dirname(os.path.dirname(
@@ -221,6 +221,7 @@ FULL_NAME_TO_PRODUCT_MAPPING = {
     "Ubuntu 20.04": "ubuntu2004",
     "WRLinux 8": "wrlinux8",
     "WRLinux 1019": "wrlinux1019",
+    "WRLinux LTS21": "wrlinuxlts21",
 }
 
 
@@ -250,7 +251,7 @@ MULTI_PLATFORM_MAPPING = {
     "multi_platform_rhv": ["rhv4"],
     "multi_platform_sle": ["sle12", "sle15"],
     "multi_platform_ubuntu": ["ubuntu1604", "ubuntu1804", "ubuntu2004"],
-    "multi_platform_wrlinux": ["wrlinux8", "wrlinux1019"],
+    "multi_platform_wrlinux": ["wrlinux8", "wrlinux1019", "wrlinuxlts21"],
 }
 
 RHEL_CENTOS_CPE_MAPPING = {
-- 
2.25.1

